src/data/padics/padic_numbers.lean

```diff
@@ -217,6 +220,56 @@ do [h, f, g] ← l.mmap tactic.i_to_expr,
 end
 
 namespace padic_seq
+
+
+section valuation
+open cau_seq
+variables {p : ℕ} [fact p.prime]
+
+/-! ### Valuation on `padic_seq` -/
+
+/--
+The `p`-adic valuation on `ℚ` lifts to `padic_seq p`.
+`valuation f` is defined to be the valuation of the (`ℚ`-valued) stationary point of `f`.
+-/
+def valuation (f : padic_seq p) : ℤ :=
+if hf : f ≈ 0 then 0 else padic_val_rat p (f (stationary_point hf))
+
+lemma norm_eq_pow_val {f : padic_seq p} (hf : ¬ f ≈ 0) :
+  f.norm = p^(-f.valuation : ℤ) :=
+begin
+  rw [norm, valuation, dif_neg hf, dif_neg hf, padic_norm, if_neg],
+  intro H,
+  apply cau_seq.not_lim_zero_of_not_congr_zero hf,
+  intros ε hε,
+  use (stationary_point hf),
+  intros n hn,
+  rw stationary_point_spec hf (le_refl _) hn,
+  simpa [H] using hε,
+end
+
+lemma val_eq_iff_norm_eq {f g : padic_seq p} (hf : ¬ f ≈ 0) (hg : ¬ g ≈ 0) :
+  f.valuation = g.valuation ↔ f.norm = g.norm :=
+begin
+  rw [norm_eq_pow_val hf, norm_eq_pow_val hg, ← neg_inj, fpow_inj],
+  { exact_mod_cast nat.prime.pos ‹_› },
+  { exact_mod_cast nat.prime.ne_one ‹_› },
+end
+
+/- lemma val_mul {f g : padic_seq p} (hf : ¬ f ≈ 0) (hg : ¬ g ≈ 0) :
+  valuation (f * g) = valuation f + valuation g :=
+begin
+  unfold valuation,
+  have := mul_not_equiv_zero hf hg,
+  split_ifs,
+  { contradiction },
+  { padic_index_simp [this, hf, hg] }
+end
+
+#check padic_val_rat.mul -/
```

Comment:
Remove dead code?

Reply:
Oops, didn't mean to include that commit...